File: service/src/com/android/car/power/SilentModeHandler.java
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.car.power;

import android.annotation.NonNull;
import android.os.FileObserver;
import android.os.SystemProperties;
import android.util.IndentingPrintWriter;
import android.util.Slog;

import com.android.car.CarLog;
import com.android.internal.annotations.GuardedBy;
import com.android.internal.annotations.VisibleForTesting;

import libcore.io.IoUtils;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Objects;

/**
 * Class to handle Silent Mode and Non-Silent Mode.
 *
 * <p>This monitors {@code /sys/power/pm_silentmode_hw_state} to figure out when to switch to Silent
 * Mode and updates {@code /sys/power/pm_silentmode_kernel} to tell early-init services about Silent
 * Mode change. Also, it handles forced Silent Mode for testing purpose, which is given through
 * reboot reason.
 */
final class SilentModeHandler {
    private static final String TAG = CarLog.TAG_POWER + "_"
            + SilentModeHandler.class.getSimpleName();

    private static final String SYSFS_FILENAME_GPIO_MONITORING =
            "/sys/power/pm_silentmode_hw_state";
    private static final String SYSFS_FILENAME_KERNEL_SILENTMODE =
            "/sys/power/pm_silentmode_kernel";
    private static final String VALUE_SILENT_MODE = "1";
    private static final String VALUE_NON_SILENT_MODE = "0";
    private static final String SYSTEM_BOOT_REASON = "sys.boot.reason";
    private static final String FORCED_NON_SILENT = "reboot,forcednonsilent";
    private static final String FORCED_SILENT = "reboot,forcedsilent";

    private final Object mLock = new Object();
    private final CarPowerManagementService mService;
    private final String mGpioMonitoringFileName;
    private final String mKernelSilentModeFileName;

    private FileObserver mFileObserver;
    @GuardedBy("mLock")
    private boolean mSilentModeByGpio;
    private boolean mForcedMode;

    SilentModeHandler(@NonNull CarPowerManagementService service) {
        this(service, SYSFS_FILENAME_GPIO_MONITORING, SYSFS_FILENAME_KERNEL_SILENTMODE,
                SystemProperties.get(SYSTEM_BOOT_REASON));
    }

    @VisibleForTesting
    SilentModeHandler(@NonNull CarPowerManagementService service,
            @NonNull String gpioMonitoringFileName, @NonNull String kernelSilentModeFileName,
            @NonNull String bootReason) {
        Objects.requireNonNull(service, "CarPowerManagementService must not be null");
        Objects.requireNonNull(gpioMonitoringFileName,
                "Filename for GPIO monitoring must not be null");
        Objects.requireNonNull(kernelSilentModeFileName,
                "Filename for Kernel silent mode must not be null");
        Objects.requireNonNull(bootReason, "Boot reason must not be null");
        mService = service;
        mGpioMonitoringFileName = gpioMonitoringFileName;
        mKernelSilentModeFileName = kernelSilentModeFileName;
        switch (bootReason) {
            case FORCED_SILENT:
                Slog.i(TAG, "Starting in forced silent mode");
                mForcedMode = true;
                mSilentModeByGpio = true;
                break;
            case FORCED_NON_SILENT:
                Slog.i(TAG, "Starting in forced non-silent mode");
                mForcedMode = true;
                mSilentModeByGpio = false;
                break;
            default:
                mForcedMode = false;
        }
    }

    void init() {
        if (mForcedMode) {
            // In forced mode, mSilentModeByGpio is set by shell command and not changed by real
            // GPIO signal. It's okay to access without lock.
            updateKernelSilentMode(mSilentModeByGpio);
            Slog.i(TAG, "Now in forced mode: monitoring " + mGpioMonitoringFileName
                    + " is disabled");
        } else {
            startMonitoringSilentModeGpio();
        }
    }

    void release() {
        if (mFileObserver != null) {
            mFileObserver.stopWatching();
            mFileObserver = null;
        }
    }

    void dump(IndentingPrintWriter writer) {
        synchronized (mLock) {
            writer.printf("Monitoring GPIO signal: %b\n", mFileObserver != null);
            writer.printf("Silent mode by GPIO signal: %b\n", mSilentModeByGpio);
            writer.printf("Forced silent mode: %b\n", mForcedMode);
        }
    }

    boolean isSilentMode() {
        synchronized (mLock) {
            return mSilentModeByGpio;
        }
    }

    void querySilentModeGpio() {
        if (mFileObserver != null) {
            mFileObserver.onEvent(FileObserver.MODIFY, mGpioMonitoringFileName);
        }
    }

    void updateKernelSilentMode(boolean silent) {
        try (BufferedWriter writer =
                new BufferedWriter(new FileWriter(mKernelSilentModeFileName))) {
            writer.write(silent ? VALUE_SILENT_MODE : VALUE_NON_SILENT_MODE);
            writer.flush();
        } catch (IOException e) {
            Slog.w(TAG, "Failed to update " + mKernelSilentModeFileName + " to "
                    + (silent ? VALUE_SILENT_MODE : VALUE_NON_SILENT_MODE));
        }
    }

    private void startMonitoringSilentModeGpio() {
        File monitorFile = new File(mGpioMonitoringFileName);
        if (!monitorFile.exists()) {
            Slog.w(TAG, "Failed to start monitoring Silent Mode GPIO: " + mGpioMonitoringFileName
                    + " doesn't exist");
            return;
        }
        mFileObserver = new FileObserver(monitorFile, FileObserver.MODIFY) {
            @Override
            public void onEvent(int event, String filename) {
                boolean newSilentMode;
                boolean oldSilentMode;
                synchronized (mLock) {
                    oldSilentMode = mSilentModeByGpio;
                    try {
                        String contents = IoUtils.readFileAsString(mGpioMonitoringFileName).trim();
                        mSilentModeByGpio = VALUE_SILENT_MODE.equals(contents);
                        Slog.i(TAG, mGpioMonitoringFileName + " indicates "
                                + (mSilentModeByGpio ? "silent" : "non-silent") + " mode");
                    } catch (Exception e) {
                        Slog.w(TAG, "Failed to read " + mGpioMonitoringFileName + ": " + e);
                        return;
                    }
                    newSilentMode = mSilentModeByGpio;
                }
                if (newSilentMode != oldSilentMode) {
                    mService.notifySilentModeChange(newSilentMode);
                }
            }
        };
        mFileObserver.startWatching();
        // Trigger the observer to get the initial contents
        querySilentModeGpio();
    }
}

File: cpp/powerpolicy/server/src/SilentModeHandler.cpp
/*
 * Copyright (c) 2021, The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#define LOG_TAG "carpowerpolicyd"

#include "SilentModeHandler.h"

#include "CarPowerPolicyServer.h"

#include <android-base/file.h>
#include <android-base/properties.h>
#include <android-base/stringprintf.h>
#include <android-base/strings.h>

#include <sys/inotify.h>
#include <sys/stat.h>
#include <unistd.h>

namespace android {
namespace frameworks {
namespace automotive {
namespace powerpolicy {

using ::android::Mutex;
using ::android::base::Error;
using ::android::base::GetProperty;
using ::android::base::ReadFileToString;
using ::android::base::Result;
using ::android::base::SetProperty;
using ::android::base::StringPrintf;
using ::android::base::Trim;
using ::android::base::unique_fd;
using ::android::base::WriteStringToFd;

namespace {

constexpr const char kPropertySystemBootReason[] = "sys.boot.reason";
constexpr const char kSilentModeHwStateFilename[] = "/sys/power/pm_silentmode_hw_state";
constexpr const char kKernelSilentModeFilename[] = "/sys/power/pm_silentmode_kernel";
// To prevent boot animation from being started.
constexpr const char kPropertyNoBootAnimation[] = "debug.sf.nobootanimation";
// To stop boot animation while it is being played.
constexpr const char kPropertyBootAnimationExit[] = "service.bootanim.exit";
constexpr int kEventBufferSize = 512;

bool fileExists(const char* filename) {
    struct stat buffer;
    return stat(filename, &buffer) == 0;
}

}  // namespace

SilentModeHandler::SilentModeHandler(ISilentModeChangeHandler* handler) :
      mSilentModeByHwState(false),
      mSilentModeHwStateFilename(kSilentModeHwStateFilename),
      mKernelSilentModeFilename(kKernelSilentModeFilename),
      mSilentModeChangeHandler(handler),
      mFdInotify(-1) {
    mBootReason = GetProperty(kPropertySystemBootReason, "");
}

void SilentModeHandler::init() {
    if (mBootReason == kBootReasonForcedSilent) {
        mForcedMode = true;
        mSilentModeByHwState = true;
    } else if (mBootReason == kBootReasonForcedNonSilent) {
        mForcedMode = true;
        mSilentModeByHwState = false;
    }
    if (mForcedMode) {
        handleSilentModeChange(mSilentModeByHwState);
        mSilentModeChangeHandler->notifySilentModeChange(mSilentModeByHwState);
        ALOGI("Now in forced mode: monitoring %s is disabled", kSilentModeHwStateFilename);
    } else {
        startMonitoringSilentModeHwState();
    }
}

void SilentModeHandler::release() {
    stopMonitoringSilentModeHwState(/*shouldWaitThread=*/false);
}

bool SilentModeHandler::isSilentMode() {
    Mutex::Autolock lock(mMutex);
    return mSilentModeByHwState;
}

void SilentModeHandler::stopMonitoringSilentModeHwState(bool shouldWaitThread) {
    if (mIsMonitoring) {
        mIsMonitoring = false;
        inotify_rm_watch(mFdInotify, mWdSilentModeHwState);
        mWdSilentModeHwState = -1;
        if (shouldWaitThread && mSilentModeMonitoringThread.joinable()) {
            mSilentModeMonitoringThread.join();
        }
    }
    mFdInotify.reset(-1);
}

Result<void> SilentModeHandler::dump(int fd, const Vector<String16>& /*args*/) {
    const char* indent = "  ";
    WriteStringToFd(StringPrintf("%sMonitoring HW state: %s\n", indent,
                                 mIsMonitoring ? "true" : "false"),
                    fd);
    WriteStringToFd(StringPrintf("%sForced silent mode: %s\n", indent,
                                 mForcedMode ? "true" : "false"),
                    fd);
    if (mIsMonitoring) {
        Mutex::Autolock lock(mMutex);
        WriteStringToFd(StringPrintf("%sSilent mode by HW state: %s\n", indent,
                                     mSilentModeByHwState ? "silent" : "non-silent"),
                        fd);
    }
    return {};
}

void SilentModeHandler::startMonitoringSilentModeHwState() {
    if (mIsMonitoring) {
        ALOGW("Silent Mode monitoring is already started");
        return;
    }
    if (mFdInotify < 0) {
        mFdInotify.reset(inotify_init1(IN_CLOEXEC));
        if (mFdInotify < 0) {
            ALOGE("Failed to start monitoring Silent Mode HW state: creating inotify instance "
                  "failed (errno = %d)",
                  errno);
            return;
        }
    }
    const char* filename = mSilentModeHwStateFilename.c_str();
    if (!fileExists(filename)) {
        ALOGW("Failed to start monitoring Silent Mode HW state: %s doesn't exist", filename);
        mFdInotify.reset(-1);
        return;
    }
    // TODO(b/178843534): Additional masks might be needed to detect sysfs change.
    const uint32_t masks = IN_MODIFY;
    mWdSilentModeHwState = inotify_add_watch(mFdInotify, filename, masks);
    mIsMonitoring = true;
    mSilentModeMonitoringThread = std::thread([this]() {
        char eventBuf[kEventBufferSize];
        struct inotify_event* event;
        constexpr size_t inotifyEventSize = sizeof(*event);
        ALOGI("Monitoring %s started", mSilentModeHwStateFilename.c_str());
        while (mIsMonitoring) {
            int eventPos = 0;
            int numBytes = read(mFdInotify, eventBuf, sizeof(eventBuf));
            if (numBytes < static_cast<int>(inotifyEventSize)) {
                if (errno == EINTR) {
                    ALOGW("System call interrupted. Wait for inotify event again.");
                    continue;
                }
                mIsMonitoring = false;
                inotify_rm_watch(mFdInotify, mWdSilentModeHwState);
                mWdSilentModeHwState = -1;
                mFdInotify.reset(-1);
                ALOGW("Failed to wait for change at %s (errno = %d)",
                      mSilentModeHwStateFilename.c_str(), errno);
                return;
            }
            while (numBytes >= static_cast<int>(inotifyEventSize)) {
                int eventSize;
                event = (struct inotify_event*)(eventBuf + eventPos);
                if (event->wd == mWdSilentModeHwState && (event->mask & masks)) {
                    handleSilentModeHwStateChange();
                }
                eventSize = inotifyEventSize + event->len;
                numBytes -= eventSize;
                eventPos += eventSize;
            }
        }
        ALOGI("Monitoring %s ended", mSilentModeHwStateFilename.c_str());
    });
    // Read the current silent mode HW state.
    handleSilentModeHwStateChange();
}

void SilentModeHandler::handleSilentModeHwStateChange() {
    if (!mIsMonitoring) {
        return;
    }
    std::string buf;
    if (!ReadFileToString(mSilentModeHwStateFilename.c_str(), &buf)) {
        ALOGW("Failed to read %s", mSilentModeHwStateFilename.c_str());
        return;
    }
    bool newSilentMode;
    bool oldSilentMode;
    {
        Mutex::Autolock lock(mMutex);
        oldSilentMode = std::exchange(mSilentModeByHwState, Trim(buf) == kValueSilentMode);
        newSilentMode = mSilentModeByHwState;
    }
    if (newSilentMode != oldSilentMode) {
        ALOGI("%s is set to %s", mSilentModeHwStateFilename.c_str(),
              newSilentMode ? "silent" : "non-silent");
        handleSilentModeChange(newSilentMode);
        mSilentModeChangeHandler->notifySilentModeChange(newSilentMode);
    }
}

void SilentModeHandler::handleSilentModeChange(bool silent) {
    if (auto ret = updateKernelSilentMode(silent); !ret.ok()) {
        ALOGW("Failed to update kernel silent mode: %s", ret.error().message().c_str());
    }
    if (auto ret = enableBootAnimation(!silent); !ret.ok()) {
        ALOGW("Failed to %s boot animation: %s", mSilentModeByHwState ? "disabling" : "enabling",
              ret.error().message().c_str());
    }
}

Result<void> SilentModeHandler::enableBootAnimation(bool enabled) {
    const std::string value = enabled ? "0" : "1";
    if (!SetProperty(kPropertyNoBootAnimation, value)) {
        return Error() << "Failed to set " << kPropertyNoBootAnimation << " property to " << value;
    }
    if (!enabled) {
        if (!SetProperty(kPropertyBootAnimationExit, value)) {
            return Error() << "Failed to set " << kPropertyBootAnimationExit << " property to "
                           << value;
        }
    }
    return {};
}

Result<void> SilentModeHandler::updateKernelSilentMode(bool silent) {
    int fd = open(mKernelSilentModeFilename.c_str(), O_WRONLY | O_NONBLOCK);
    if (fd < 0) {
        return Error() << "Failed to open " << mKernelSilentModeFilename;
    }
    Result<void> status = {};
    if (const auto& value = silent ? kValueSilentMode : kValueNonSilentMode;
        !WriteStringToFd(value, fd)) {
        status = Error() << "Failed to write " << value << " to fd " << fd;
    }
    close(fd);
    return status;
}

}  // namespace powerpolicy
}  // namespace automotive
}  // namespace frameworks
}  // namespace android
